// argos-cli/lib/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _path = _interopRequireDefault(require("path"));

var _convict = _interopRequireDefault(require("convict"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const config = (0, _convict.default)({
  endpoint: {
    doc: 'Argos API endpoint',
    format: 'url',
    default: 'https://api.argos-ci.com',
    env: 'ARGOS_API_ENDPOINT'
  },
  commit: {
    doc: 'Git commit',
    format: String,
    default: '',
    env: 'ARGOS_COMMIT'
  },
  branch: {
    doc: 'Git branch',
    format: String,
    default: '',
    env: 'ARGOS_BRANCH'
  },
  token: {
    doc: 'Repository token',
    format: String,
    default: '',
    env: 'ARGOS_TOKEN'
  },
  externalBuildId: {
    doc: 'External build id (batch mode)',
    format: String,
    default: '',
    env: 'ARGOS_EXTERNAL_BUILD_ID'
  },
  batchCount: {
    doc: 'Batch count expected (batch mode)',
    format: String,
    default: '',
    env: 'ARGOS_BATCH_COUNT'
  }
});
const NODE_ENV = process.env.NODE_ENV || 'development';

if (NODE_ENV !== 'production') {
  config.loadFile(_path.default.join(__dirname, `../config/${NODE_ENV}.json`));
}

config.validate();
var _default = config;
exports.default = _default;
